src/client/java/teammates/client/scripts/DataMigrationBaseScript.java

```diff
@@ -0,0 +1,91 @@
+package teammates.client.scripts;
+
+import java.util.List;
+
+import teammates.client.remoteapi.RemoteApiClient;
+import teammates.client.scripts.util.LoopHelper;
+import teammates.common.datatransfer.attributes.EntityAttributes;
+
+/**
+ * Base script to be used as template for all data migration scripts.
+ *
+ * @param <T> The entity type to be migrated by the script.
+ */
+public abstract class DataMigrationBaseScript<T extends EntityAttributes<?>> extends RemoteApiClient {
+
+    /**
+     * If true, the script will not perform actual data migration.
+     */
+    protected abstract boolean isPreview();
+
+    /**
+     * Gets all the entities that are to be filtered for migration.
+     */
+    protected abstract List<T> getEntities();
+
+    /**
+     * Checks whether data migration is needed.
+     */
+    protected abstract boolean isMigrationNeeded(T entity);
+
+    /**
+     * Prints information necessary for previewing the migration process.
+     */
+    protected abstract void printPreviewInformation(T entity);
+
+    /**
+     * Migrates the entity.
+     */
+    protected abstract void migrate(T entity) throws Exception;
+
+    /**
+     * Performs any post-migration action.
+     *
+     * <p>This can also be used for batch migration, i.e. in that case {@link #migrate(EntityAttributes)}
+     * will only be used to queue up the entities that need to be migrated.
+     */
+    protected abstract void postAction();
```

Comment:
Can we have a default method that does nothing? E.g. https://github.com/TEAMMATES/teammates/blob/19a0e8ee0e7d6a9e609de972257060fdee2d15f9/src/main/java/teammates/common/util/retry/Retryable.java#L38-L44

Reply:
I have no strong opinion of either option, but this option has the downside that the implementing classes of `DataMigrationBaseScript` may not even be aware of the existence of this procedure without looking at the base class.

Not sure how it will affect future scripts as currently there is only one script that actually makes use of this method. Similarly for `printPreviewInformation`.